summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala

```diff
@@ -21,43 +21,65 @@ import scala.reflect.ClassTag
-   */
-  private def tickConfig = {
-    val boltConfig = new BacktypeStormConfig
-    boltConfig.put(BacktypeStormConfig.TOPOLOGY_TICK_TUPLE_FREQ_SECS, java.lang.Integer.valueOf(1))
-    boltConfig
-  }
+  private def registerNodes: Topology =
+    stormDag.nodes.foldLeft(Topology.EMPTY) {
+      case (topology, node: SummerNode[Storm]) =>
+        topology.withBolt(getNodeName(node), registerSummerBolt(node))._2
+      case (topology, node: FlatMapNode[Storm]) =>
+        topology.withBolt(getNodeName(node), FlatMapBoltProvider(this, node).apply)._2
+      case (topology, node: SourceNode[Storm]) =>
+        topology.withSpout(getNodeName(node), SpoutProvider(this, node).apply)._2
+    }
 
-  private def scheduleFlatMapper(node: FlatMapNode[Storm], topologyBuilder: TopologyBuilder) = {
-    val bolt: BaseBolt[Any, Any] = FlatMapBoltProvider(this, node).apply
+  private def registerEdges(topologyWithNodes: Topology): Topology =
+    stormDag.nodes.foldLeft(topologyWithNodes) {
+      case (topology, node: SummerNode[Storm]) =>
+        val edgeType = EdgeType.AggregatedKeyValues[Any, Any](getSummerKeyValueShards(node))
+        registerIncomingEdges(topology, node, edgeType)
+      case (topology, node: FlatMapNode[Storm]) =>
+        registerIncomingEdges(topology, node, flatMapEdgeType(node))
+      case (topology, node: SourceNode[Storm]) => topology
+        // ignore, no incoming edges into sources
+    }
 
-    val parallelism = getOrElse(node, DEFAULT_FM_PARALLELISM).parHint
-    val declarer = topologyBuilder.setBolt(getNodeName(node), bolt, parallelism).addConfigurations(tickConfig)
-    bolt.applyGroupings(declarer)
+  private def flatMapEdgeType(node: FlatMapNode[Storm]): EdgeType[_] = {
+    val usePreferLocalDependency = getOrElse(node, DEFAULT_FM_PREFER_LOCAL_DEPENDENCY)
+    logger.info(s"[${getNodeName(node)}] usePreferLocalDependency: ${usePreferLocalDependency.get}")
+    if (usePreferLocalDependency.get) {
+      EdgeType.itemWithLocalOrShuffleGrouping[Any]
+    } else {
+      EdgeType.itemWithShuffleGrouping[Any]
+    }
   }
 
-  private def scheduleSpout(node: SourceNode[Storm], topologyBuilder: TopologyBuilder) = {
-    val nodeName = stormDag.getNodeName(node)
-    val (sourceParalleism, stormSpout) = SpoutProvider(this, node).apply
-    topologyBuilder.setSpout(nodeName, stormSpout, sourceParalleism)
+  private def registerIncomingEdges(topology: Topology, node: StormNode, edgeType: EdgeType[_]): Topology = {
+    val nodeId: ReceivingId[Any] = getId(node).asInstanceOf[Topology.ReceivingId[Any]]
+    stormDag.dependenciesOf(node).foldLeft(topology) { (current, upstreamNode) =>
+      current.withEdge(Topology.Edge[Any](
+        getId(upstreamNode),
+        edgeType.asInstanceOf[EdgeType[Any]],
+        nodeId))
+    }
   }
 
-  private def scheduleSummerBolt(node: SummerNode[Storm], topologyBuilder: TopologyBuilder): Unit = {
+  private def getId(node: StormNode): Topology.EmittingId[_] = node match {
```

Comment:
I did more or less on purpose %) But I think it's ok to be consistent here as well